fairlearn/grid_search.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+This module implements a 1d grid search for binary classification with a single
+binary protected attribute.
+
+This is a simplification of the full exponentiated gradient search
+and produces a sequence of models, each of which corresponds to a
+different accuracy/fairness trade-off. The dimensionality of the search
+increases with the number of possible values for the protected attribute,
+so this technique does not scale beyond a binary protected attribute.
+"""
+
+import copy
+import numpy as np
+
+class BinaryClassificationGridSearch:
+    def __init__(self):
+        # Nothing to do
+        return
+
+    def _generate_p0_p1(self, y):
+        """ Function to compute p0 and p1 for the given
+        set of labels
+        """
+        unique, counts = np.unique(y, return_counts=True)
+
+        if len(unique) != 2:
+            raise RuntimeError("Supplied y not binary")
+
+        if not np.array_equal(unique, [0, 1]):
+            raise RuntimeError("Supplied y labels not 0 or 1")
+
+        p0 = counts[0] / len(y)
+        p1 = counts[1] / len(y)
+
+        return p0, p1
+
+    def _generate_weights(self, y, protected_attribute, L, p_ratio):
+        weights = []
+
+        for current_y, current_a in zip(y, protected_attribute):
+            w = 1e128
+            if current_a == 0:
+                w = 2*current_y - 1 - L * p_ratio
+            else:
+                w = 2*current_y - 1 + L
+
+            weights.append(w)
+
+        return np.array(weights)
+
```

Comment:
Should we use a vectorized version rather than an iteration? My guess is it's going to be more concise--likely easier to read.

Reply:
Changed. It at least gets rid of that horrible append.